grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/ReasonerUtils.java

```diff
@@ -287,38 +298,70 @@ public static void printAnswers(Set<Map<String, Concept>> answers) {
     }
 
     /**
-     * convert a given role type to a set of relation types in which it can appear including  the role type hierarchy
+     * convert a given role type to a map of relation types in which it can play roles and the corresponding role types including entity type hierarchy
      */
-    public static final Function<RoleType, Set<RelationType>> roleToRelationTypes =
-            role -> role.subTypes().stream()
-                    .flatMap(r -> r.relationTypes().stream())
-                    .filter(rt -> !rt.isImplicit())
-                    .collect(toSet());
+    public static final Function<RoleType, Multimap<RelationType, RoleType>> roleHierarchyToRelationTypesWithRoles =
```

Comment:
Just a suggestion but I would make an interface in these cases. A function to a multimap is a bit difficult to read.